pycbc/waveform/spa_tmplt.py

```diff
@@ -211,14 +221,25 @@ def spa_tmplt(**kwds):
         f_lower = kwds['f_lower']
         delta_f = kwds['delta_f']
         kmin = int(f_lower / float(delta_f))
-        vISCO = 1. / sqrt(6.)
-        fISCO = vISCO * vISCO * vISCO / piM
-
-        if 'f_upper' in kwds:
-            fISCO = kwds['f_upper']
 
-        kmax = int(fISCO / delta_f)
-        f_max = ceilpow2(fISCO)
+        # Get max frequency one way or another
+        # f_final is assigned default value 0 in parameters.py
+        if 'f_final' in kwds and kwds['f_final'] > 0.:
+            fmax = kwds['f_final']
+        elif 'f_upper' in kwds:
+            fmax = kwds['f_upper']
+            warnings.warn('f_upper is deprecated in favour of f_final !',
+                          DeprecationWarning)
+        else:
+            # Schwarzschild ISCO frequency
+            vISCO = 1. / sqrt(6.)
+            fmax = vISCO * vISCO * vISCO / piM
+        if fmax <= f_lower:
+            raise NoWaveformError("cannot generate waveform: f_lower >= "
```

Comment:
Looks like you cannot import the waveform module here because that creates a circular import. Just use a ValueError or RuntimeError, I think they are good enough here.

Also, I find having two variables named `f_max` and `fmax` confusing. I would just use `f_max` all the way.

Reply:
Ah, ugh, that is a bit unpleasant.  I guess the whole package would have to be restructured to allow these specialized errors to be used by every module, or they would have to be put in __init__.py (?)